### ISSUE TITLE
Softmax on f16/bf16 is numerically broken on long axes (rows sum to ~2.0, not 1.0)

### ISSUE THREAD

**Issue by v-code01** [huggingface/candle#3727]
### Severity

Silent correctness bug (wrong numerical output, no error/panic) on the **attention hot path**. Affects any f16/bf16 model doing softmax over a long axis (e.g. attention over long sequences). The CPU result also disagrees with the Metal backend.

### Summary

`candle_nn::ops::softmax_last_dim` runs the whole computation in the tensor's dtype. For `f16`/`bf16` the exp-denominator is summed sequentially in the narrow dtype, which **saturates** on a long axis: an f16 running sum of ones stalls at 2048 (once the sum's ULP exceeds 1, each `+1` rounds back). So the row is normalized by 2048 instead of its true length, and the "probabilities" no longer sum to 1.

### Repro (CPU, no GPU needed)

```rust
use candle_core::{Device, Tensor, DType};
use candle_nn::ops::softmax_last_dim;

fn main() -> candle_core::Result<()> {
    // 4096 equal logits in f16 => every exp = 1, denominator should be 4096.
    let t = Tensor::zeros((1, 4096), DType::F16, &Device::Cpu)?;
    let sm = softmax_last_dim(&t)?;
    // Measure in f32 so the check itself doesn't saturate.
    let row_sum = sm.to_dtype(DType::F32)?.sum_all()?.to_scalar::<f32>()?;
    let first  = sm.to_dtype(DType::F32)?.flatten_all()?.get(0)?.to_scalar::<f32>()?;
    println!("row_sum = {row_sum}   first_prob = {first}");
    Ok(())
}
```

### Expected vs actual

```
Expected: row_sum = 1.0        first_prob = 1/4096 = 0.000244
Actual:   row_sum = 2.0        first_prob = 1/2048 = 0.000488
```

Every probability is ~2x too large; the softmax is not normalized. bf16 is worse (8-bit mantissa saturates even sooner). f32/f64 are fine.

### Root cause

The `f16`/`bf16` path sums the exponentials in the narrow dtype. The Metal kernel already keeps the softmax denominator in f32 (`MD::d` is `float`), so the backends disagree.

### Fix

Compute the f16/bf16 path in f32 (max, exp, denominator, normalization) and cast back, matching the Metal kernel; f32/f64 paths unchanged. I've opened #3717 with this fix + regression tests. After it, an f16 softmax over a length-4096 axis matches the Metal backend to max-abs 1.2e-7 (both rows sum to 1.0).

### Environment

candle main; reproduces on CPU (Apple M4, macOS), backend-independent.

**Comment by RajaBabu15** [comment]
I'll work on this — fixing f16/bf16 softmax saturation on long axes by accumulating in f32 (with regression test).

**Comment by RajaBabu15** [comment]
Update: looks like **#3717** already implements this exact fix (f16/bf16 softmax via f32 accumulation + regression tests) and is open/mergeable. I'll step back on this issue to avoid a duplicate PR — happy to review #3717 instead.

Pivoting to another open correctness issue.

**Comment by RajaBabu15** [comment]
For anyone landing here: the fix for this issue is already in open PR **#3717** (not my branch). I claimed this earlier by mistake and stepped back — see above.

**Comment by v-code01** [comment]
Thanks @RajaBabu15, appreciate you taking a look and the pointer to #3717 🙏

Quick note in case it's useful: #3717 now also folds in two sibling instances of the same root cause (CPU accumulating f16/bf16 reductions in the narrow dtype) — `sum`/`mean` (`vec_reduce_sum`) and `avg_pool2d` — since they saturate the same way on long axes / large windows and disagree with the Metal backend. All three are behind regression tests and the f32/f64 paths are untouched. A review would be very welcome if you have the bandwidth.
